istioctl/cmd/webhook.go

```diff
@@ -0,0 +1,510 @@
+
+import (
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"io/ioutil"
+	"time"
+
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/spf13/cobra"
+
+	"k8s.io/api/admissionregistration/v1beta1"
+
+	"github.com/ghodss/yaml"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	caCertServiceAccount          = "default"
+	caCertServiceAccountNamespace = "default"
+	caKeyInK8sSecret              = "ca.crt"
+	certKeyInK8sSecret            = "cert-chain.pem"
+)
+
+type enableCliOptions struct {
+	// Whether enable the webhook configuration of Galley
+	enableValidationWebhook bool
+	// The local file path of webhook CA certificate
+	caCertPath string
+	// The file path of the webhook configuration.
+	webhookConfigPath string
+	// The name of the ValidatingWebhookConfiguration to manage
+	validatingWebhookConfigName string
+	// The service name of the validating webhook to manage
+	validatingWebhookServiceName string
+	// Max time (in seconds) for checking the validating webhook.
+	// If the validating webhook is not ready in the given time, exit.
+	// Otherwise, apply the webhook configuration.
+	maxTimeForCheckingWebhookServer int
+	// Max time (in seconds) for waiting the webhook certificate to be readable.
+	maxTimeForReadingWebhookCert int
+	// The name of a webhook secret in the cluster.
+	// istioctl will check whether the webhook certificate
+	// in the secret is issued by the certificate in the k8s service account or the given
+	// CA cert, if any.
+	webhookSecretName string
+	// The namespace of a webhook secret in the cluster.
+	webhookSecretNameSpace string
+}
+
+type disableCliOptions struct {
+	// Whether disable the webhook configuration of Galley
+	disableValidationWebhook bool
+	// The name of the ValidatingWebhookConfiguration to manage
+	validatingWebhookConfigName string
+}
+
+type statusCliOptions struct {
+	// Whether display the webhook configuration of Galley
+	validationWebhook bool
+	// The name of the ValidatingWebhookConfiguration to manage
+	validatingWebhookConfigName string
+}
+
+// Webhook command to manage webhook configurations
+func Webhook() *cobra.Command {
+	var (
+		enableOpts  = enableCliOptions{}
+		disableOpts = disableCliOptions{}
+		statusOpts  = statusCliOptions{}
+	)
+
+	cmd := &cobra.Command{
+		Use:   "webhook",
+		Short: "webhook command to manage webhook configurations",
+	}
+
+	cmd.AddCommand(newEnableCmd(&enableOpts))
+	cmd.AddCommand(newDisableCmd(&disableOpts))
+	cmd.AddCommand(newStatusCmd(&statusOpts))
+
+	return cmd
+}
+
+func newEnableCmd(opts *enableCliOptions) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "enable",
+		Short: "Enable webhook configurations",
+		Example: `
+# Enable the webhook configuration of Galley with the given webhook configuration
+istioctl experimental post-install webhook enable --validation --webhook-secret-name istio.webhook.galley 
+    --webhook-secret-namespace istio-system --namespace istio-system --config-path /etc/validatingwebhookconfiguration.yaml
```

Comment:
Agree, I will make --validation and --injection flags be mutually exclusive. An user needs to run two commands for Galley and Sidecar Injector, respectively, rather than one long unwiedly command.

Reply:
Assuming we default the webhook/secret names, the following is the minimum users will need to type to have a usable install.

```bash
istioctl x post-install webhook enable --validation \
    --config-path install/kubernetes/helm/istio/charts/galley/validatingwebhookconfiguration.yaml
istioctl x post-install webhook enable --injection \
    --config-path install/kubernetes/helm/istio/charts/sidecar-injector/mutatingwebhookconfiguration.yaml
```

This still seems too much. It would be great to have something more like the following. This might mean adding back the per-webhook flags for advanced use cases. cc @costinm for his 2c.

```bash
$ istioctl x post-install webhooks enable --help
usage: istioctl x post-install webhooks enable [--no-injection] [--no-validation] 

# enable everything
$ istioctl x post-install webhooks enable 

# enable everything except injection
$ istioctl x post-install webhooks enable --no-injection
```

Reply:
Based on offline discussions, the command will look like: 
istioctl x post-install webhook enable [--validation] [--injection] [--validation-secret=<default> [--injection-secret=<default>] --config-path [--ca-bundle-file=""] [--namespace=<default>].

If validation and injection webhooks are installed in different namespaces, the user calls two CLI commands, each with different --namespace.

Reply:
Done.